### PR TITLE
Time out typing over federation

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/federation/federation_client.py [hunk 1]
```diff
@@ -136,9 +136,7 @@ def send_edu(self, destination, edu_type, content, key=None):
 
         sent_edus_counter.inc()
 
-        # TODO, add errback, etc.
         self._transaction_queue.enqueue_edu(edu, key=key)
-        return defer.succeed(None)
```
Comment (richvdh) [inline synapse/federation/federation_client.py]: is this related, or just a cleanup?

Comment (erikjohnston) [inline synapse/federation/federation_client.py]: Just a clean up

File: synapse/handlers/typing.py [hunk 2]
```diff
@@ -85,32 +123,31 @@ def started_typing(self, target_user, auth_user, room_id, timeout):
...
-            self.clock.cancel_call_later(self._member_typing_timer[member])
+        now = self.clock.time_msec()
+        self._member_typing_until[member] = now + timeout
 
-        def _cb():
-            logger.debug(
-                "%s has timed out in %s", target_user.to_string(), room_id
-            )
-            self._stopped_typing(member)
+        self.wheel_timer.insert(
```
Comment (richvdh) [inline synapse/handlers/typing.py]: is there a danger of these stacking up, so that we get lots of entries for the same user?

Comment (richvdh) [inline synapse/handlers/typing.py]: I guess there is, but it's harmless?

File: synapse/handlers/typing.py [hunk 3]
```diff
@@ -85,32 +123,31 @@ def started_typing(self, target_user, auth_user, room_id, timeout):
...
+            now=now,
+            obj=member,
+            then=now + timeout,
+        )
 
-        self._member_typing_until[member] = until
-        self._member_typing_timer[member] = self.clock.call_later(
-            timeout / 1000.0, _cb
+        self.wheel_timer.insert(
```
Comment (richvdh) [inline synapse/handlers/typing.py]: could we skip this if `was_present` is true?

File: synapse/handlers/typing.py [hunk 4]
```diff
@@ -85,32 +123,31 @@ def started_typing(self, target_user, auth_user, room_id, timeout):
...
+            now=now,
+            obj=member,
+            then=now + FEDERATION_PING_INTERVAL,
```
Comment (richvdh) [inline synapse/handlers/typing.py]: I'm slightly wondering if the code would be clearer if we had separate `WheelTimer` instances for timing out the typing, and doing the federation pings.

File: synapse/federation/federation_client.py (hunk 1 above)
Comment (richvdh) [inline synapse/federation/federation_client.py]: ack

File: synapse/handlers/typing.py [hunk 5]
```diff
@@ -53,20 +62,49 @@ def __init__(self, hs):
...
-        to avoid a "Reactor was unclean" warning."""
-        for t in self._member_typing_timer.values():
-            self.clock.cancel_call_later(t)
+        self.clock.looping_call(
+            self._handle_timeouts,
+            5000,
+        )
+
+    def _handle_timeouts(self):
+        logger.info("Handling typing timeout")
```
Comment (erikjohnston) [inline synapse/handlers/typing.py]: True, will change.

File: synapse/handlers/typing.py (hunk 2 above)
Comment (erikjohnston) [inline synapse/handlers/typing.py]: Yeah they'll stack up, but the `_handle_timeouts` is designed not care about spurious wake ups.

File: synapse/handlers/typing.py (hunk 4 above)
Comment (erikjohnston) [inline synapse/handlers/typing.py]: Hmm, maybe, but I quite like the style of having a single function that occasionally gets woken up and checks if there is anything it needs to do.

**Comment by erikjohnston** [comment]
@matrixbot retest this please
